src/render/function.rs

```diff
@@ -0,0 +1,300 @@
+            format!(
+                "Could not find theme {theme:?}, must be one of {}",
+                ts.themes
+                    .keys()
+                    .map(|s| format!("{s:?}"))
+                    .collect::<Vec<_>>()
+                    .join(", ")
+            )
+        })?,
+    );
+
+    let threads = (threads == 0)
+        .then(num_cpus::get)
+        .unwrap_or(threads)
+        .clamp(1, num_cpus::get());
+    let (mut line_num, longest_line_chars, background) = if threads < 2 {
+        let mut line_num: u32 = 0;
+        let mut longest_line_chars = 0;
+        let mut background = None;
+        let mut highlighter = cache.new_plain_highlighter();
+        for (file_index, ((path, content), num_content_lines, _lines_so_far)) in
+            content.into_iter().enumerate()
+        {
+            progress.inc();
+            if should_interrupt.load(Ordering::Relaxed) {
+                bail!("Cancelled by user")
+            }
+            if !plain {
+                if let Some(hl) = cache.highlighter_for_file_name(path)? {
+                    highlighter = hl;
+                }
+            }
+
+            if display_to_be_processed_file {
+                progress.info(format!("{path:?}"))
+            }
+            let out = chunk::process(
+                content,
+                &mut img,
+                |line| highlighter.highlight_line(line, ss),
+                chunk::Context {
+                    column_width,
+                    line_height,
+                    total_line_count,
+                    highlight_truncated_lines,
+                    line_num,
+                    lines_per_column,
+                    fg_color,
+                    bg_color,
+                    file_index,
+                    color_modulation,
+                },
+            )?;
+            longest_line_chars = out.longest_line_in_chars.max(longest_line_chars);
+            line_num += num_content_lines as u32;
+            line_progress.inc_by(num_content_lines);
+            background = out.background;
+        }
+
+        (line_num, longest_line_chars, background)
+    } else {
+        // multi-threaded rendering overview:
+        //
+        // Spawns threadpool and each file to be renered is sent to a thread as a message via a flume channel.
+        // Upon recieving a message, a thread renders the entire file to an image of one column width.
+        // and then returns that image to this main thread via a flume channel, to be stitched together
+        // into one large image. The ordering of files rendered in the final image is remembered and
+        // independant of thread rendering order.
+
+        let mut line_num: u32 = 0;
+        let mut longest_line_chars = 0;
+        let mut background = None;
+        let file_index = AtomicUsize::default();
+        std::thread::scope(|scope| -> anyhow::Result<()> {
+            let (ttx, trx) = flume::unbounded();
```

Comment:
I like not sending content to scoped threads via a channel, and instead letting the scoped threads just read the content, and keeping track of which one to render with an `AtomicUsize`. I don't remember how much overhead channels have. Do you think this will increase performance?

Reply:
I like to think of channels as: use them, unless you want to send millions of things per second, as I do equate them with a fancy lock which does slow things down if it's contended. 
Here we are talking 100k images being sent in huge repositories, and that's no issue at all especially because the time it takes to compute the work is considerable, so the sending is spread out over minutes.

As general rule, I use what makes the code easy to understand and optimize later once there is something to measure.